particle-iot/device-os
# [Gen3] USB bugfixes

### Problem

1. USB serial port state is not being reliably detected on Gen 3 devices. This easily causes a deadlock when writing data into USB serial port when there is no other side to read out the data
2. USB control requests fail sometimes partially due to ^^, there is also another issue whether extra control requests are forwarded

### Solution

1. Refactors USB implementation for Gen 3. Utilize SOF for detecting that the other side is not reading data out and consider the connection to the host broken. This is similar (but not quite) to how we do things on Gen 2. The difference is that we are not marking the serial port is closed and simply abort TX transfer and flush 
2. Filter out non-vendor requests

This PR also changes the implementation of the critical sections for Nordic SDK to use BASEPRI. This change supersedes and fixes the issue mentioned in #2091.

### Steps to Test

1. Make the device generate some data on Serial e.g. by using SerialLogHandler. Open the serial port, close it, wait for a while for the device to generate some more data. Attempt to run a USB control request - it should not timeout.
2. Validate the test from https://github.com/particle-iot/device-os/pull/2091

### Example App

N/A

### References

- [CH42073]
- [CH50902]

---

### Completeness
